bftengine/src/bcstatetransfer/RVBManager.cpp

```diff
@@ -95,15 +107,21 @@ void RVBManager::init(bool fetching) {
     }
   }
 
-  if (!loaded_from_data_store && (desc.maxBlockId > 0)) {
+  // Do not attempt to build RVT in case
+  // 1. RVT is already build from RVB data stored in last checkpoint
+  // 2. Max Block Id from last stored checkpoint is 0
+  // 3. last reachable block == default genesis block (eg. 1)
+
+  if ((!loaded_from_data_store) && ((desc.maxBlockId) || (last_reachable_block_num > 1))) {
     uint64_t num_rvbs_added{};
     // If desc data is valid, try to reconstruct by reading digests from storage (no persistency data was found)
     DurationTracker<std::chrono::milliseconds> reconstruct_dt("reconstruct_dt", true);
     auto genesis_block_id = as_->getGenesisBlockNum();
     LOG_INFO(logger_, "Reconstructing RVB data" << KVLOG(loaded_from_data_store, desc.maxBlockId, genesis_block_id));
-    if (genesis_block_id <= desc.maxBlockId) {
-      num_rvbs_added =
-          addRvbDataOnBlockRange(genesis_block_id, desc.maxBlockId, std::optional<Digest>(desc.digestOfMaxBlockId));
+    if ((genesis_block_id <= desc.maxBlockId) || (desc.maxBlockId <= last_reachable_block_num)) {
+      num_rvbs_added = addRvbDataOnBlockRange(genesis_block_id,
+                                              std::max(desc.maxBlockId, last_reachable_block_num),
+                                              std::optional<Digest>(desc.digestOfMaxBlockId));
```

Comment:
I'm worried about the distributed correctness of all the above, in rare cases (and we are now dealing with rare cases).
You should give priority always to maxBlockId since it relays on a checkpoint and the checkpoint should be the same for all replicas, hence all init and have the same tree.
constructing from last_reachable_block_num is very tricky, we might get for some time a non-identical RVT, if there was some human error or things we can think about now. We are building a special tree here, not relying on checkpoints. This is very sensitive if the network is standing still immediately after cloning.
If and ONLY if desc.maxBlockId is 0, then you should build the tree according to last_reachable_block_num.
What you did here is a combination of both, by getting the max between both ( last_reachable_block_num will be larger if both are non 0).
I might e wrong here, need to think more..  Let's discuss this offline.

Reply:
Thanks. Addressed.